File: Lab6-RailwayPlanning/Python Solution/edge.py

class Edge:
    """Class describing edge of graph in lab6"""
    def __init__(self, dest1, dest2, cap):
        self.destination1 = dest1  # these are node objects
        self.destination2 = dest2
        self.capacity = cap
        self.flow_front = 0
        self.flow_back = 0

    def is_full(self, node):
        """Check if flow=capacity towards node 'node'"""
        if node == self.destination2:
            return self.capacity == self.flow_front
        else:
            return self.capacity == self.flow_back

    def get_capacity(self, node):
        """Get capacity towards node 'node'"""
        if node == self.destination2:
            return self.capacity - self.flow_front
        else:
            return self.capacity - self.flow_back

    def update_flow(self, node, delta):
        """Increase flow with 'delta' towards node 'node'"""
        if node == self.destination2:
            self.flow_front += delta
            self.flow_back -= delta
        else:
            self.flow_back += delta
            self.flow_front -= delta

    def print_edge(self):
        output = "Edge: from " + str(self.destination1) + " to " + str(self.destination2) \
                 + " with capacity " + str(self.capacity) + ", flow_front " + str(self.flow_front) \
                 + " and flow_back " + str(self.flow_back)
        print(output)

File: Lab6-RailwayPlanning/C++ Solution/edge.cc
#include "edge.h"
#include "node.h"

using std::cout;
using std::endl;

Edge::Edge(Node* n1, Node* n2, const int& cap) : destination1(n1), destination2(n2), capacity(cap),
                                            front_flow(0), back_flow(0) {}

/**
 * Get capacity of edge towards node 'n'
 */ 
int Edge::get_capacity(const Node* n){
    if(n == destination2)
        return capacity - front_flow;
    else return capacity - back_flow;
}

/**
 * Increase flow towards node 'n' with 'delta'
 */
void Edge::update_flow(const Node* n, const int& delta){
    if(n == destination2){
        front_flow += delta;
        back_flow -= delta; 
    } else {
        back_flow += delta;
        front_flow -= delta;
    }
}

/**
 * Checks if flow = capacity towards node 'n', returns true if so, otherwise false
 */ 
bool Edge::is_full(const Node* n){
    if(n == destination2)
        return capacity == front_flow;
    else return capacity == back_flow;
}

void Edge::print_edge(){
    cout << "Edge from " << destination1->index << " to " << destination2->index << 
            " with capacity " << capacity << ", front flow " << front_flow <<
            " and back flow " << back_flow;
}